cas-server-support-yubikey/src/main/java/org/jasig/cas/adaptors/yubikey/YubiKeyAuthenticationHandler.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Licensed to Apereo under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work
+ * for additional information regarding copyright ownership.
+ * Apereo licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License.  You may obtain a
+ * copy of the License at the following location:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.jasig.cas.adaptors.yubikey;
+
+import com.yubico.client.v2.ResponseStatus;
+import com.yubico.client.v2.VerificationResponse;
+import com.yubico.client.v2.YubicoClient;
+import com.yubico.client.v2.exceptions.YubicoValidationFailure;
+import com.yubico.client.v2.exceptions.YubicoVerificationException;
+import org.jasig.cas.authentication.HandlerResult;
+import org.jasig.cas.authentication.PreventedException;
+import org.jasig.cas.authentication.UsernamePasswordCredential;
+import org.jasig.cas.authentication.handler.support.AbstractUsernamePasswordAuthenticationHandler;
+import org.springframework.beans.factory.InitializingBean;
+
+import javax.annotation.PostConstruct;
+import javax.security.auth.login.AccountNotFoundException;
+import javax.security.auth.login.FailedLoginException;
+import javax.validation.constraints.NotNull;
+import java.security.GeneralSecurityException;
+
+/**
+ * An authentication handler that uses the Yubico cloud validation
+ * platform to authenticate one-time password tokens that are
+ * issued by a YubiKey device. To use YubiCloud you need a
+ * client id and an API key which must be obtained from Yubico.
+ *
+ * <p>For more info, please visit
+ * <a href="http://yubico.github.io/yubico-java-client/">this link</a></p>
+ *
+ * @author Misagh Moayyed
+ * @since 4.1
+ */
+public class YubiKeyAuthenticationHandler extends AbstractUsernamePasswordAuthenticationHandler
```

Comment:
I specifically wanted to avoid creating a new credential type. That would have meant way more work to configure the webflow, the UI, have separate resovlers, etc. Instead, this is modeled as uid/password, where you actually submit your userid as the uid, and your password is in fact the one-time password generated by Yubikey. We are still using username/password authN. The format of the password is different a bit.